File: MusicEventManagementSystem/MusicEventManagementSystem.Client/src/frontend/performer-negotations/services/contractService.ts
import { api } from '../../shared/services/apiService';

// Utility functions for date conversion
const formatDateForApi = (date: string | Date | null | undefined): string | null => {
  if (!date) return null;
  
  // If it's already a string in the correct format, return it
  if (typeof date === 'string') {
    // Check if it's already in ISO format
    if (date.includes('T') && date.includes('Z')) {
      return date;
    }
    // Otherwise parse and format it
    const parsedDate = new Date(date);
    return isNaN(parsedDate.getTime()) ? null : parsedDate.toISOString();
  }
  
  // If it's a Date object
  if (date instanceof Date) {
    return isNaN(date.getTime()) ? null : date.toISOString();
  }
  
  return null;
};

// Process UpdateContractDto to ensure proper date formatting
const processUpdateDto = (dto: UpdateContractDto): UpdateContractDto => {
  const processed = { ...dto };
  
  // Convert date fields to proper ISO strings or undefined
  if (dto.signedAt !== undefined) {
    const formatted = formatDateForApi(dto.signedAt);
    processed.signedAt = formatted || undefined;
  }
  if (dto.finalVersionDate !== undefined) {
    const formatted = formatDateForApi(dto.finalVersionDate);
    processed.finalVersionDate = formatted || undefined;
  }
  if (dto.depositDueDate !== undefined) {
    const formatted = formatDateForApi(dto.depositDueDate);
    processed.depositDueDate = formatted || undefined;
  }
  if (dto.finalPaymentDueDate !== undefined) {
    const formatted = formatDateForApi(dto.finalPaymentDueDate);
    processed.finalPaymentDueDate = formatted || undefined;
  }
  if (dto.stakeholderReviewDate !== undefined) {
    const formatted = formatDateForApi(dto.stakeholderReviewDate);
    processed.stakeholderReviewDate = formatted || undefined;
  }
  
  return processed;
};

// Process CreateContractDto to ensure proper date formatting
const processCreateDto = (dto: CreateContractDto): CreateContractDto => {
  const processed = { ...dto };
  
  // Convert date fields to proper ISO strings if they exist
  if (dto.depositDueDate) {
    const formatted = formatDateForApi(dto.depositDueDate);
    processed.depositDueDate = formatted || undefined;
  }
  if (dto.finalPaymentDueDate) {
    const formatted = formatDateForApi(dto.finalPaymentDueDate);
    processed.finalPaymentDueDate = formatted || undefined;
  }
  
  return processed;
};

// Simplified types matching new backend DTOs
export interface ContractDto {
  contractId: number;
  title: string;
  contractType: string;
  price: number;
  version: string;
  status: string;
  createdAt: string;
  signedAt?: string;

  // Contract Document Information
  contractFilePath: string;
  finalVersionDate?: string;

  // Requirements
  technicalRequirements: string;
  accommodationRequirements: string;

  // Payment Information
  depositAmount?: number;
  finalPaymentAmount?: number;
  depositDueDate?: string;
  finalPaymentDueDate?: string;
  paymentMethod: string;
  isDepositPaid: boolean;
  isFinalPaymentPaid: boolean;

  // Banking Information
  bankName: string;
  bankAccountNumber: string;
  bankRoutingNumber: string;
  bankAccountHolderName: string;
  bankIBAN: string;
  bankSWIFT: string;

  // Review Information
  reviewedByStakeholders: boolean;
  stakeholderReviewDate?: string;

  // Notes
  notes: string;

  // Related entities
  performerId: number;
  performerName?: string;
  eventId?: number;
  eventTitle?: string;
  eventLocation?: string;
  eventDate?: string;
}

export interface UpdateContractDto {
  title?: string;
  contractType?: string;
  price?: number;
  version?: string;
  status?: string;
  signedAt?: string;

  // Contract Document Information
  contractFilePath?: string;
  finalVersionDate?: string;

  // Requirements
  technicalRequirements?: string;
  accommodationRequirements?: string;

  // Payment Information
  depositAmount?: number;
  finalPaymentAmount?: number;
  depositDueDate?: string;
  finalPaymentDueDate?: string;
  paymentMethod?: string;
  isDepositPaid?: boolean;
  isFinalPaymentPaid?: boolean;

  // Banking Information
  bankName?: string;
  bankAccountNumber?: string;
  bankRoutingNumber?: string;
  bankAccountHolderName?: string;
  bankIBAN?: string;
  bankSWIFT?: string;

  // Review Information
  reviewedByStakeholders?: boolean;
  stakeholderReviewDate?: string;

  // Notes
  notes?: string;
}

export interface CreateContractDto {
  title: string;
  contractType: string;
  price: number;
  version: string;
  status: string;

  // Contract Document Information
  contractFilePath?: string;

  // Requirements
  technicalRequirements?: string;
  accommodationRequirements?: string;

  // Payment Information
  depositAmount?: number;
  finalPaymentAmount?: number;
  depositDueDate?: string;
  finalPaymentDueDate?: string;
  paymentMethod?: string;

  // Banking Information
  bankName?: string;
  bankAccountNumber?: string;
  bankRoutingNumber?: string;
  bankAccountHolderName?: string;
  bankIBAN?: string;
  bankSWIFT?: string;

  // Notes
  notes?: string;

  // Foreign Keys
  performerId: number;
  eventId?: number;
}
const API_ENDPOINT = '/contract';

export const contractService = {
  // Get all contracts
  getAllContracts: async (): Promise<ContractDto[]> => {
    try {
      const response = await api.get<ContractDto[]>(API_ENDPOINT);
      return response.data;
    } catch (error) {
      console.error('Error fetching contracts:', error);
      throw error;
    }
  },

  // Get contract by ID
  getContractById: async (id: number): Promise<ContractDto> => {
    try {
      const response = await api.get<ContractDto>(`${API_ENDPOINT}/${id}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching contract:', error);
      throw error;
    }
  },

  // Create new contract
  createContract: async (contract: CreateContractDto): Promise<ContractDto> => {
    try {
      // Process the contract data to ensure proper date formatting
      const processedContract = processCreateDto(contract);
      const response = await api.post<ContractDto>(API_ENDPOINT, processedContract);
      return response.data;
    } catch (error) {
      console.error('Error creating contract:', error);
      throw error;
    }
  },

  // Update contract (simplified - no workflow stages)
  updateContract: async (id: number, contract: UpdateContractDto): Promise<ContractDto> => {
    try {
      // Process the contract data to ensure proper date formatting
      const processedContract = processUpdateDto(contract);
      const response = await api.put<ContractDto>(`${API_ENDPOINT}/${id}`, processedContract);
      return response.data;
    } catch (error) {
      console.error('Error updating contract:', error);
      throw error;
    }
  },

  // Delete contract
  deleteContract: async (id: number): Promise<void> => {
    try {
      await api.delete(`${API_ENDPOINT}/${id}`);
    } catch (error) {
      console.error('Error deleting contract:', error);
      throw error;
    }
  },

  // Create contract draft from negotiation (only available in phases 3, 4, 5)
  createContractDraft: async (negotiationId: number): Promise<ContractDto> => {
    try {
      const response = await api.post<ContractDto>(`${API_ENDPOINT}/draft/${negotiationId}`);
      return response.data;
    } catch (error) {
      console.error('Error creating contract draft:', error);
      throw error;
    }
  },

  // Get contract with details including performer and event data
  getContractWithDetails: async (contractId: number): Promise<ContractDto> => {
    try {
      const response = await api.get<ContractDto>(`${API_ENDPOINT}/${contractId}/details`);
      return response.data;
    } catch (error) {
      console.error('Error fetching contract details:', error);
      throw error;
    }
  },

  // Get contracts by negotiation ID
  getContractsByNegotiation: async (negotiationId: number): Promise<ContractDto[]> => {
    try {
      const response = await api.get<ContractDto[]>(`${API_ENDPOINT}/negotiation/${negotiationId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching contracts by negotiation:', error);
      throw error;
    }
  }
};

File: MusicEventManagementSystem/MusicEventManagementSystem.API/Services/ContractService.cs
using MusicEventManagementSystem.API.Models;
using MusicEventManagementSystem.API.Repositories.IRepositories;
using MusicEventManagementSystem.API.Services.IService;
using MusicEventManagementSystem.API.DTOs;

namespace MusicEventManagementSystem.API.Services
{
    public class ContractService : IContractService
    {
        private readonly IContractRepository _contractRepository;
        private readonly INegotiationRepository _negotiationRepository;
        private readonly INegotiationPhaseRepository _negotiationPhaseRepository;

        public ContractService(
            IContractRepository contractRepository,
            INegotiationRepository negotiationRepository,
            INegotiationPhaseRepository negotiationPhaseRepository)
        {
            _contractRepository = contractRepository;
            _negotiationRepository = negotiationRepository;
            _negotiationPhaseRepository = negotiationPhaseRepository;
        }

        public async Task<IEnumerable<Contract>> GetAllContractsAsync()
        {
            return await _contractRepository.GetAllAsync();
        }

        public async Task<Contract?> GetContractByIdAsync(int id)
        {
            return await _contractRepository.GetByIdAsync(id);
        }

        public async Task<Contract> CreateContractAsync(Contract contract)
        {
            contract.CreatedAt = DateTime.UtcNow;
            await _contractRepository.AddAsync(contract);
            await _contractRepository.SaveChangesAsync();
            return contract;
        }

        public async Task<bool> DeleteContractAsync(int id)
        {
            var contract = await _contractRepository.GetByIdAsync(id);
            if (contract == null)
            {
                return false;
            }

            _contractRepository.Delete(contract);
            await _contractRepository.SaveChangesAsync();
            return true;
        }

        public async Task<ContractDto> CreateContractDraftFromNegotiationAsync(int negotiationId)
        {
            try
            {
                Console.WriteLine($"DEBUG: Starting contract creation for negotiation {negotiationId}");
                
                // Get negotiation with performer and event details
                var negotiation = await _negotiationRepository.GetNegotiationWithDetailsAsync(negotiationId);
                if (negotiation == null)
                {
                    Console.WriteLine($"DEBUG: Negotiation {negotiationId} not found");
                    throw new ArgumentException($"Negotiation with ID {negotiationId} not found.");
                }

                Console.WriteLine($"DEBUG: Found negotiation {negotiationId}, performer: {negotiation.Performer?.Name}, event: {negotiation.Event?.Name}");

                // Get current phase for debugging
                var currentPhase = await _negotiationPhaseRepository.GetCurrentNegotiationPhaseAsync(negotiationId);
                Console.WriteLine($"DEBUG: Current phase: {currentPhase?.PhaseId} - {currentPhase?.Phase?.PhaseName}");
                
                // Temporary: Allow contract creation in any phase for testing
                // TODO: Re-enable phase validation once testing is complete
                /*
                // Validate phase - only allow contract draft creation in phases 3, 4, or 5
                if (currentPhase == null || !IsPhaseAllowedForContractEditing(currentPhase.PhaseId))
                {
                    throw new InvalidOperationException($"Contract can only be created/edited in Contract Negotiations (3), Contract Draft (4), or Final Agreement (5) phases. Current phase: {currentPhase?.PhaseId}");
                }
                */

                // Create contract draft with both performer and event information
                var contract = new Contract
                {
                    Title = $"Performance Contract - {negotiation.Performer?.Name ?? "Performer"} at {negotiation.Event?.Name ?? "Event"}",
                    ContractType = "Performance",
                    Price = negotiation.ProposedFee,
                    Version = "1.0",
                    Status = "Draft",
                    CreatedAt = DateTime.UtcNow,
                    PerformerId = negotiation.PerformerId,
                    EventId = negotiation.EventId,
                    // Initialize banking fields to avoid database issues
                    BankName = "",
                    BankAccountNumber = "",
                    BankRoutingNumber = "",
                    BankAccountHolderName = "",
                    BankIBAN = "",
                    BankSWIFT = ""
                };

                Console.WriteLine($"DEBUG: About to save contract to database");
                await _contractRepository.AddAsync(contract);
                await _contractRepository.SaveChangesAsync();
                
                Console.WriteLine($"DEBUG: Contract saved successfully with ID: {contract.ContractId}");
                
                // Reload with navigation properties
                var savedContract = await _contractRepository.GetByIdAsync(contract.ContractId);
                return MapToContractDto(savedContract!);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"DEBUG: Error creating contract: {ex.Message}");
                Console.WriteLine($"DEBUG: Stack trace: {ex.StackTrace}");
                throw;
            }
        }

        public async Task<ContractDto?> GetContractWithDetailsAsync(int contractId)
        {
            var contract = await _contractRepository.GetByIdAsync(contractId);
            if (contract == null) return null;

            return MapToContractDto(contract);
        }

        public async Task<ContractDto?> UpdateContractAsync(int contractId, UpdateContractDto updateDto)
        {
            var existingContract = await _contractRepository.GetByIdAsync(contractId);
            if (existingContract == null) return null;

            // TODO: Add phase validation - for now allow editing to test functionality
            // In production, you would check the current negotiation phase here

            // Update only provided fields (null means don't update)
            if (updateDto.Title != null) existingContract.Title = updateDto.Title;
            if (updateDto.ContractType != null) existingContract.ContractType = updateDto.ContractType;
            if (updateDto.Price.HasValue) existingContract.Price = updateDto.Price.Value;
            if (updateDto.Version != null) existingContract.Version = updateDto.Version;
            if (updateDto.Status != null) existingContract.Status = updateDto.Status;
            if (updateDto.SignedAt.HasValue) existingContract.SignedAt = updateDto.SignedAt;

            // Document information
            if (updateDto.ContractFilePath != null) existingContract.ContractFilePath = updateDto.ContractFilePath;
            if (updateDto.FinalVersionDate.HasValue) existingContract.FinalVersionDate = updateDto.FinalVersionDate;

            // Requirements
            if (updateDto.TechnicalRequirements != null) existingContract.TechnicalRequirements = updateDto.TechnicalRequirements;
            if (updateDto.AccommodationRequirements != null) existingContract.AccommodationRequirements = updateDto.AccommodationRequirements;

            // Payment information
            if (updateDto.DepositAmount.HasValue) existingContract.DepositAmount = updateDto.DepositAmount;
            if (updateDto.FinalPaymentAmount.HasValue) existingContract.FinalPaymentAmount = updateDto.FinalPaymentAmount;
            if (updateDto.DepositDueDate.HasValue) existingContract.DepositDueDate = updateDto.DepositDueDate;
            if (updateDto.FinalPaymentDueDate.HasValue) existingContract.FinalPaymentDueDate = updateDto.FinalPaymentDueDate;
            if (updateDto.PaymentMethod != null) existingContract.PaymentMethod = updateDto.PaymentMethod;
            if (updateDto.IsDepositPaid.HasValue) existingContract.IsDepositPaid = updateDto.IsDepositPaid.Value;
            if (updateDto.IsFinalPaymentPaid.HasValue) existingContract.IsFinalPaymentPaid = updateDto.IsFinalPaymentPaid.Value;

            // Banking information
            if (updateDto.BankName != null) existingContract.BankName = updateDto.BankName;
            if (updateDto.BankAccountNumber != null) existingContract.BankAccountNumber = updateDto.BankAccountNumber;
            if (updateDto.BankRoutingNumber != null) existingContract.BankRoutingNumber = updateDto.BankRoutingNumber;
            if (updateDto.BankAccountHolderName != null) existingContract.BankAccountHolderName = updateDto.BankAccountHolderName;
            if (updateDto.BankIBAN != null) existingContract.BankIBAN = updateDto.BankIBAN;
            if (updateDto.BankSWIFT != null) existingContract.BankSWIFT = updateDto.BankSWIFT;

            // Review information
            if (updateDto.ReviewedByStakeholders.HasValue) existingContract.ReviewedByStakeholders = updateDto.ReviewedByStakeholders.Value;
            if (updateDto.StakeholderReviewDate.HasValue) existingContract.StakeholderReviewDate = updateDto.StakeholderReviewDate;

            // Notes
            if (updateDto.Notes != null) existingContract.Notes = updateDto.Notes;

            _contractRepository.Update(existingContract);
            await _contractRepository.SaveChangesAsync();

            return MapToContractDto(existingContract);
        }

        private static bool IsPhaseAllowedForContractEditing(int phaseId)
        {
            // Phase 3: Contract Negotiations
            // Phase 4: Contract Draft  
            // Phase 5: Final Agreement
            return phaseId == 3 || phaseId == 4 || phaseId == 5;
        }

        // Helper method to map Contract to ContractDto
        private static ContractDto MapToContractDto(Contract contract)
        {
            return new ContractDto
            {
                ContractId = contract.ContractId,
                Title = contract.Title,
                ContractType = contract.ContractType,
                Price = contract.Price,
                Version = contract.Version,
                Status = contract.Status,
                CreatedAt = contract.CreatedAt,
                SignedAt = contract.SignedAt,

                // Contract Document Information
                ContractFilePath = contract.ContractFilePath,
                FinalVersionDate = contract.FinalVersionDate,

                // Requirements
                TechnicalRequirements = contract.TechnicalRequirements,
                AccommodationRequirements = contract.AccommodationRequirements,

                // Payment Information
                DepositAmount = contract.DepositAmount,
                FinalPaymentAmount = contract.FinalPaymentAmount,
                DepositDueDate = contract.DepositDueDate,
                FinalPaymentDueDate = contract.FinalPaymentDueDate,
                PaymentMethod = contract.PaymentMethod,
                IsDepositPaid = contract.IsDepositPaid,
                IsFinalPaymentPaid = contract.IsFinalPaymentPaid,

                // Banking Information
                BankName = contract.BankName,
                BankAccountNumber = contract.BankAccountNumber,
                BankRoutingNumber = contract.BankRoutingNumber,
                BankAccountHolderName = contract.BankAccountHolderName,
                BankIBAN = contract.BankIBAN,
                BankSWIFT = contract.BankSWIFT,

                // Review Information
                ReviewedByStakeholders = contract.ReviewedByStakeholders,
                StakeholderReviewDate = contract.StakeholderReviewDate,

                // Notes
                Notes = contract.Notes,

                // Related entities (populated from navigation properties)
                PerformerId = contract.PerformerId,
                PerformerName = contract.Performer?.Name,
                EventId = contract.EventId,
                EventTitle = contract.Event?.Name,
                EventDate = contract.Event?.Interval,
                EventLocation = contract.Event?.Location?.Name ?? (contract.Event?.LocationId != null ? $"Location ID: {contract.Event.LocationId}" : null)
            };
        }

        public async Task<IEnumerable<ContractDto>> GetContractsByNegotiationAsync(int negotiationId)
        {
            try
            {
                // Get negotiation to find performer and event IDs
                var negotiation = await _negotiationRepository.GetNegotiationWithDetailsAsync(negotiationId);
                if (negotiation == null)
                {
                    return new List<ContractDto>();
                }

                // Get all contracts that match this negotiation's performer and event
                var allContracts = await _contractRepository.GetAllAsync();
                var matchingContracts = allContracts
                    .Where(c => c.PerformerId == negotiation.PerformerId && c.EventId == negotiation.EventId)
                    .ToList();

                return matchingContracts.Select(MapToContractDto);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"DEBUG: Error getting contracts for negotiation {negotiationId}: {ex.Message}");
                throw;
            }
        }
    }
}
